crates/accelerate/src/synthesis/linear/lnn.rs

```diff
@@ -0,0 +1,217 @@
+        gates.push((
+            StandardGate::CXGate,
+            smallvec![],
+            smallvec![Qubit((2 * i) as u32), Qubit((2 * i + 1) as u32)],
+        ))
+    }
+
+    for i in 0..((n + 1) / 2 - 1) {
+        gates.push((
+            StandardGate::CXGate,
+            smallvec![],
+            smallvec![Qubit((2 * i + 2) as u32), Qubit((2 * i + 1) as u32)],
+        ))
+    }
+}
+
+// A single layer of CX gates.
+fn _append_cx_stage2(gates: &mut LnnGatesVec, n: usize) {
+    for i in 0..(n / 2) {
+        gates.push((
+            StandardGate::CXGate,
+            smallvec![],
+            smallvec![Qubit((2 * i + 1) as u32), Qubit((2 * i) as u32)],
+        ))
+    }
+
+    for i in 0..((n + 1) / 2 - 1) {
+        gates.push((
+            StandardGate::CXGate,
+            smallvec![],
+            smallvec![Qubit((2 * i + 1) as u32), Qubit((2 * i + 2) as u32)],
+        ))
+    }
+}
+
+// Appends correct phase gate during CZ synthesis
+fn _append_phase_gate(pat_val: usize, gates: &mut LnnGatesVec, qubit: usize) {
+    // Add phase gates: s, sdg or z
+    if pat_val % 4 == 1 {
+        gates.push((
+            StandardGate::SdgGate,
+            smallvec![],
+            smallvec![Qubit(qubit as u32)],
+        ))
+    } else if pat_val % 4 == 2 {
+        gates.push((
+            StandardGate::ZGate,
+            smallvec![],
+            smallvec![Qubit(qubit as u32)],
+        ))
+    } else if pat_val % 4 == 3 {
+        gates.push((
+            StandardGate::SGate,
+            smallvec![],
+            smallvec![Qubit(qubit as u32)],
+        ))
+    }
+}
+
+// Synthesis of a CZ circuit for linear nearest neighbor (LNN) connectivity,
+// based on Maslov and Roetteler.
+pub(super) fn synth_cz_depth_line_mr(matrix: ArrayView2<bool>) -> (usize, LnnGatesVec) {
+    let num_qubits = matrix.raw_dim()[0];
+    let pats = _create_patterns(num_qubits as isize);
+
+    // s_gates[i] = 0, 1, 2 or 3 for a gate id, sdg, z or s on qubit i respectively
+    let mut s_gates = Array1::<usize>::zeros(num_qubits);
+
+    let mut patlist: Vec<(isize, isize)> = Vec::new();
+
+    let mut gates = LnnGatesVec::new();
+
+    for i in 0..num_qubits {
+        for j in (i + 1)..num_qubits {
+            if matrix[[i, j]] {
+                // CZ(i,j) gate
+                s_gates[[i]] += 2; // qc.z[i]
+                s_gates[[j]] += 2; // qc.z[j]
+                patlist.push((i as isize, j as isize - 1));
+                patlist.push((i as isize, j as isize));
+                patlist.push((i as isize + 1, j as isize - 1));
+                patlist.push((i as isize + 1, j as isize));
+            }
+        }
+    }
+
+    for i in 0..((num_qubits + 1) / 2) {
+        for j in 0..num_qubits {
+            let pat_val = pats[&(i as isize, j as isize)];
+            if patlist.contains(&pat_val) {
+                let patcnt = patlist.iter().filter(|val| **val == pat_val).count();
+                s_gates[[j]] += patcnt; // qc.sdg[j]
```

Comment:
If I understand `patcnt` is either 0 or 1 -- could you add a comment explaining that this checks whether an Sdg gate must be added? I think future devs would be grateful (in the Python code it was a bit more obvious IMO) 🙂